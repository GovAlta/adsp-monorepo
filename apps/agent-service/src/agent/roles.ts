export enum ServiceRoles {
  AgentUser = 'agent-user',
  AgentTool = 'agent-tool',
}
