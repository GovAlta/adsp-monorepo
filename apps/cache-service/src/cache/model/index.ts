export * from './cacheTarget';
export * from './accessCacheTarget';
export { accessApiId } from './access';
