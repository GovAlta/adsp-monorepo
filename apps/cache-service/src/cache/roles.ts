export enum ServiceRoles {
  CacheReader = 'cache-reader',
}
