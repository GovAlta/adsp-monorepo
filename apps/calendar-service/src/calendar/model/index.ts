export * from './calendar';
export * from './event';
