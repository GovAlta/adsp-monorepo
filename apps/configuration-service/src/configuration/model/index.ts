export * from './configuration';
export * from './activeRevisionEntity';
