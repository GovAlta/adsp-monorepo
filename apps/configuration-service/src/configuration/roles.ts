export enum ConfigurationServiceRoles {
  Reader = 'configuration-reader',
  ConfigurationAdmin = 'configuration-admin',
  ConfiguredService = 'configured-service',
}

export enum DirectoryServiceRoles {
  ResourceResolver = 'urn:ads:platform:directory-service:resource-resolver',
}

export enum ExportServiceRoles {
  ExportJob = 'urn:ads:platform:export-service:export-job',
}
