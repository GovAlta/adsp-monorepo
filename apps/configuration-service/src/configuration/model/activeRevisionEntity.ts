import { AdspId, isAllowedUser, UnauthorizedUserError, User } from '@abgov/adsp-service-sdk';
import { InvalidOperationError, Results, ValidationService, createLogger } from '@core-services/core-common';
import { ConfigurationRepository, ActiveRevisionRepository } from '../repository';
import { ConfigurationServiceRoles } from '../roles';
import { ConfigurationRevision, Configuration, RevisionCriteria, ActiveRevision } from '../types';
import type { Logger } from 'winston';
import { environment } from '../../environments/environment';

/**
 * Represents an aggregate context for configuration revisions.
 *
 * @export
 * @class ServiceConfigurationEntity
 * @implements {ActiveRevision}Banff, Alberta
 * @template C
 */
export class ActiveRevisionEntity implements ActiveRevision {
  constructor(
    public namespace: string,
    public name: string,
    public repository: ActiveRevisionRepository,
    public active: number,
    public tenantId?: AdspId,
    private logger?: Logger
  ) {
    if (!namespace || !name) {
      throw new InvalidOperationError('Configuration must have a namespace and name.');
    }

    if (namespace.includes(':') || name.includes(':')) {
      throw new InvalidOperationError(`Configuration and namespace and name cannot contain ':'.`);
    }

    if (!logger) {
      logger = createLogger('active-revision-modal', environment.LOG_LEVEL);
    }
  }

  public canAccess(user: User): boolean {
    return isAllowedUser(
      user,
      this.tenantId,
      [
        ConfigurationServiceRoles.ConfiguredService,
        ConfigurationServiceRoles.ConfigurationAdmin,
        ConfigurationServiceRoles.Reader,
      ],
      true
    );
  }

  public canModify(user: User): boolean {
    return isAllowedUser(
      user,
      this.tenantId,
      [ConfigurationServiceRoles.ConfiguredService, ConfigurationServiceRoles.ConfigurationAdmin],
      true
    );
  }

  public async setActiveRevision(user: User, active: number): Promise<ActiveRevision> {
    if (!this.canModify(user)) {
      throw new UnauthorizedUserError('modify configuration', user);
    }

    const activeRevision = await this.repository.setActiveRevision(this, active);

    this.active = activeRevision.active;
    return this;
  }
}
