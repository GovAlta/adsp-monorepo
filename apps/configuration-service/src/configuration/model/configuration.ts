import { AdspId, isAllowedUser, UnauthorizedUserError, User } from '@abgov/adsp-service-sdk';
import { InvalidOperationError, Results, ValidationService } from '@core-services/core-common';
import { ConfigurationRepository, ActiveRevisionRepository } from '../repository';
import { ConfigurationServiceRoles } from '../roles';
import { ConfigurationRevision, Configuration, RevisionCriteria } from '../types';
import type { Logger } from 'winston';

/**
 * Represents an aggregate context for configuration revisions.
 *
 * @export
 * @class ServiceConfigurationEntity
 * @implements {Configuration<C>}
 * @template C
 */
export class ConfigurationEntity<C = Record<string, unknown>> implements Configuration<C> {
  constructor(
    public namespace: string,
    public name: string,
    private logger: Logger,
    public repository: ConfigurationRepository,
    private activeRevisionRepository: ActiveRevisionRepository,
    public validationService: ValidationService,
    public latest?: ConfigurationRevision<C>,
    public tenantId?: AdspId,
    private schema?: Record<string, unknown>,
    public active?: number
  ) {
    if (!namespace || !name) {
      throw new InvalidOperationError('Configuration must have a namespace and name.');
    }

    if (namespace.includes(':') || name.includes(':')) {
      throw new InvalidOperationError(`Configuration and namespace and name cannot contain ':'.`);
    }

    try {
      validationService.setSchema(this.getSchemaKey(), schema);
      return;
    } catch {
      this.logger.warn(`JSON schema of ${namespace}:${name} is invalid. An empty JSON schema {} will be used.`);
    }

    /**
      Initialize the ConfigurationEntity with empty schema if schema is mis-configured or not provided.
    */
    try {
      validationService.setSchema(this.getSchemaKey(), {});
    } catch (err) {
      throw new InvalidOperationError(`Failed in adding empty schema {}: ${err.message}`);
    }
  }

  public canAccess(user: User): boolean {
    return isAllowedUser(
      user,
      this.tenantId,
      [
        ConfigurationServiceRoles.ConfiguredService,
        ConfigurationServiceRoles.ConfigurationAdmin,
        ConfigurationServiceRoles.Reader,
      ],
      true
    );
  }

  public canModify(user: User): boolean {
    return isAllowedUser(user, this.tenantId, ConfigurationServiceRoles.ConfigurationAdmin, true);
  }

  public canRegister(user: User): boolean {
    return isAllowedUser(user, this.tenantId, ConfigurationServiceRoles.ConfiguredService, true);
  }

  public mergeUpdate(update: Partial<C>): C {
    return Object.entries(update).reduce(
      (config, [key, value]) => {
        // If current value plus update value are both objects, then combine them with spread operator.
        if (typeof config[key] === 'object' && typeof value === 'object') {
          value = { ...config[key], ...value };
        }

        config[key] = value;
        return config;
      },
      { ...this.latest?.configuration }
    );
  }

  public async update(user: User, configuration: C): Promise<ConfigurationEntity<C>> {
    if (!this.canModify(user) && !this.canRegister(user)) {
      throw new UnauthorizedUserError('modify configuration', user);
    }

    if (!configuration) {
      throw new InvalidOperationError('Configuration must have a value.');
    }

    this.validationService.validate(
      `configuration '${this.namespace}:${this.name}'`,
      this.getSchemaKey(),
      configuration
    );

    const revision: ConfigurationRevision<C> = {
      revision: this.latest?.revision || 0,
      lastUpdated: new Date(),
      created: this.latest?.created || new Date(),
      configuration: configuration,
    };

    this.latest = await this.repository.saveRevision(this, revision);
    return this;
  }

  public async createRevision(user: User): Promise<ConfigurationEntity<C>> {
    if (!this.canModify(user)) {
      throw new UnauthorizedUserError('modify configuration', user);
    }
    const newRevision: ConfigurationRevision<C> = {
      revision: this.latest ? this.latest.revision + 1 : 0,
      created: new Date(),
      lastUpdated: new Date(),
      configuration: this.latest?.configuration || ({} as C),
    };

    this.latest = await this.repository.saveRevision(this, newRevision);
    return this;
  }

  public async getRevisions(
    top = 10,
    after: string = null,
    criteria: RevisionCriteria = null
  ): Promise<Results<ConfigurationRevision<C>>> {
    return this.repository.getRevisions(this, top, after, criteria);
  }

  private getSchemaKey(): string {
    return `${this.namespace}:${this.name}`;
  }

  public async setActiveRevision(user: User, active: number): Promise<ConfigurationEntity<C>> {
    if (!this.canModify(user)) {
      throw new UnauthorizedUserError('modify configuration', user);
    }

    this.active = (await this.activeRevisionRepository.setActiveRevision(this, active)).active;

    return this;
  }
}
