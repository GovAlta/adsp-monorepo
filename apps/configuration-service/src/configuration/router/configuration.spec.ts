import { adspId, User } from '@abgov/adsp-service-sdk';
import { InvalidOperationError } from '@core-services/core-common';
import { Request, Response } from 'express';
import { Logger } from 'winston';
import { ConfigurationEntity } from '../model';
import { ConfigurationServiceRoles } from '../roles';
import {
  createConfigurationRouter,
  createConfigurationRevision,
  getConfiguration,
  getConfigurationEntity,
  patchConfigurationRevision,
  getRevisions,
} from './configuration';

describe('router', () => {
  const configurationServiceId = adspId`urn:ads:platform:configuration-service`;
  const namespace = 'platform';
  const name = 'test-service';
  const tenantId = adspId`urn:ads:platform:tenant-service:v2:/tenants/test`;

  const loggerMock = {
    debug: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
  } as unknown;

  const eventServiceMock = {
    send: jest.fn(),
  };

  const validationMock = {
    setSchema: jest.fn(),
    validate: jest.fn(),
  };

  const repositoryMock = {
    get: jest.fn(),
    getRevisions: jest.fn(),
    saveRevision: jest.fn(),
  };

  beforeEach(() => {
    repositoryMock.get.mockClear();
    eventServiceMock.send.mockClear();
  });

  describe('createConfigurationRouter', () => {
    it('can create router', () => {
      const router = createConfigurationRouter({
        isConnected: () => true,
        serviceId: adspId`urn:ads:platform:configuration-service`,
        eventService: eventServiceMock,
        logger: loggerMock as Logger,
        configuration: repositoryMock,
      });

      expect(router).toBeTruthy();
    });
  });

  describe('getServiceConfigurationEntity', () => {
    it('can create handler', () => {
      const handler = getConfigurationEntity(configurationServiceId, repositoryMock);
      expect(handler).toBeTruthy();
    });

    it('can get entity', (done) => {
      const handler = getConfigurationEntity(configurationServiceId, repositoryMock, () => true);

      // Configuration definition retrieval.
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock
        )
      );
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock,
          null,
          tenantId
        )
      );

      const entity = new ConfigurationEntity(namespace, name, repositoryMock, validationMock, null, tenantId);
      repositoryMock.get.mockResolvedValueOnce(entity);

      const req = ({
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
      } as unknown) as Request;

      handler(req, null, () => {
        expect(req['entity']).toBe(entity);
        done();
      });
    });

    it('can get tenant entity', (done) => {
      const handler = getConfigurationEntity(configurationServiceId, repositoryMock, () => false);

      // Configuration definition retrieval.
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock
        )
      );
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock,
          null,
          tenantId
        )
      );

      const entity = new ConfigurationEntity(namespace, name, repositoryMock, validationMock, null, tenantId);
      repositoryMock.get.mockResolvedValueOnce(entity);

      const req = ({
        user: { isCore: true, roles: [ConfigurationServiceRoles.Reader] } as User,
        params: { namespace, name },
        query: { tenantId: tenantId.toString() },
      } as unknown) as Request;

      handler(req, null, () => {
        expect(req['entity']).toBe(entity);
        expect(repositoryMock.get.mock.calls[2][2].toString()).toEqual(tenantId.toString());
        done();
      });
    });

    it('can return error for unauthorized', (done) => {
      const handler = getConfigurationEntity(configurationServiceId, repositoryMock, () => false);

      // Configuration definition retrieval.
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock
        )
      );
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock,
          null,
          tenantId
        )
      );

      const entity = new ConfigurationEntity(namespace, name, repositoryMock, validationMock, null, tenantId);
      repositoryMock.get.mockResolvedValueOnce(entity);

      const req = ({
        user: { isCore: false, roles: [], tenantId } as User,
        params: { namespace, name },
        query: {},
      } as unknown) as Request;

      handler(req, null, (err) => {
        expect(err).toBeTruthy();
        done();
      });
    });

    it('can get entity with core definition', (done) => {
      const handler = getConfigurationEntity(configurationServiceId, repositoryMock, () => false);

      // Configuration definition retrieval.
      const configurationSchema = {};
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock,
          {
            revision: 1,
            configuration: {
              [`${namespace}:${name}`]: { configurationSchema },
            },
          }
        )
      );

      const entity = new ConfigurationEntity(namespace, name, repositoryMock, validationMock, null, tenantId);
      repositoryMock.get.mockResolvedValueOnce(entity);

      const req = ({
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
      } as unknown) as Request;

      handler(req, null, () => {
        expect(req['entity']).toBe(entity);
        expect(repositoryMock.get.mock.calls[1][3]).toBe(configurationSchema);
        done();
      });
    });

    it('can get entity with tenant definition', (done) => {
      const handler = getConfigurationEntity(configurationServiceId, repositoryMock, () => false);

      // Configuration definition retrieval.
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock
        )
      );
      const configurationSchema = {};
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock,
          {
            revision: 1,
            configuration: {
              [`${namespace}:${name}`]: { configurationSchema },
              [namespace]: { configurationSchema: {} },
            },
          },
          tenantId
        )
      );

      const entity = new ConfigurationEntity(namespace, name, repositoryMock, validationMock, null, tenantId);
      repositoryMock.get.mockResolvedValueOnce(entity);

      const req = ({
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
      } as unknown) as Request;

      handler(req, null, () => {
        expect(req['entity']).toBe(entity);
        expect(repositoryMock.get.mock.calls[2][3]).toBe(configurationSchema);
        done();
      });
    });

    it('can get entity with tenant namespace definition', (done) => {
      const handler = getConfigurationEntity(configurationServiceId, repositoryMock, () => false);

      // Configuration definition retrieval.
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock
        )
      );
      const configurationSchema = {};
      repositoryMock.get.mockResolvedValueOnce(
        new ConfigurationEntity(
          configurationServiceId.namespace,
          configurationServiceId.service,
          repositoryMock,
          validationMock,
          {
            revision: 1,
            configuration: {
              [namespace]: { configurationSchema },
            },
          },
          tenantId
        )
      );

      const entity = new ConfigurationEntity(namespace, name, repositoryMock, validationMock, null, tenantId);
      repositoryMock.get.mockResolvedValueOnce(entity);

      const req = ({
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
      } as unknown) as Request;

      handler(req, null, () => {
        expect(req['entity']).toBe(entity);
        expect(repositoryMock.get.mock.calls[2][3]).toBe(configurationSchema);
        done();
      });
    });
  });

  describe('getServiceConfiguration', () => {
    it('can create handler', () => {
      const handler = getConfiguration();
      expect(handler).toBeTruthy();
    });

    it('can get configuration', () => {
      const handler = getConfiguration();

      const entity = new ConfigurationEntity(namespace, name, repositoryMock, validationMock, null, tenantId);

      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      handler(req, (res as unknown) as Response, jest.fn());
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ latest: null, namespace, name }));
    });
  });

  describe('patchServiceConfigurationRevision', () => {
    it('can create handler', () => {
      const handler = patchConfigurationRevision(eventServiceMock);
      expect(handler).toBeTruthy();
    });

    it('can update', async () => {
      const handler = patchConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        update: jest.fn(() => Promise.resolve(entity)),
        latest: { revision: 1, configuration: { old: 'old' } },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          operation: 'UPDATE',
          update: {
            value: 'value',
          },
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ namespace, name, latest: entity.latest }));
      expect(entity.update).toHaveBeenCalledWith(req.user, expect.objectContaining({ ...req.body.update, old: 'old' }));
      expect(eventServiceMock.send).toHaveBeenCalledTimes(1);
    });

    it('can handle no existing revision on update', async () => {
      const handler = patchConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        update: jest.fn(() => Promise.resolve(entity)),
        latest: null,
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          operation: 'UPDATE',
          update: {
            value: 'value',
          },
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ namespace, name, latest: entity.latest }));
      expect(entity.update).toHaveBeenCalledWith(req.user, expect.objectContaining(req.body.update));
      expect(eventServiceMock.send).toHaveBeenCalledTimes(1);
    });

    it('can return error for update missing value', async () => {
      const handler = patchConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        update: jest.fn(() => Promise.resolve(entity)),
        latest: { revision: 1, configuration: { old: 'old' } },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          operation: 'UPDATE',
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(InvalidOperationError));
      expect(eventServiceMock.send).toHaveBeenCalledTimes(0);
    });

    it('can replace', async () => {
      const handler = patchConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        update: jest.fn(() => Promise.resolve(entity)),
        latest: { revision: 1, configuration: { old: 'old' } },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          operation: 'REPLACE',
          configuration: {
            value: 'value',
          },
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ namespace, name, latest: entity.latest }));
      expect(entity.update).toHaveBeenCalledWith(
        req.user,
        expect.objectContaining<{ value: string }>(req.body.configuration)
      );
      expect(eventServiceMock.send).toHaveBeenCalledTimes(1);
    });

    it('can return error for replace without value', async () => {
      const handler = patchConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        update: jest.fn(() => Promise.resolve(entity)),
        latest: { revision: 1, configuration: { old: 'old' } },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          operation: 'REPLACE',
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(InvalidOperationError));
      expect(eventServiceMock.send).toHaveBeenCalledTimes(0);
    });

    it('can delete', async () => {
      const handler = patchConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        update: jest.fn(() => Promise.resolve(entity)),
        latest: { revision: 1, configuration: { old: 'old' } },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          operation: 'DELETE',
          property: 'old',
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ namespace, name, latest: entity.latest }));
      expect(entity.update).toHaveBeenCalledWith(req.user, expect.not.objectContaining({ old: 'old' }));
      expect(eventServiceMock.send).toHaveBeenCalledTimes(1);
    });

    it('can handle no existing revision on delete', async () => {
      const handler = patchConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        update: jest.fn(() => Promise.resolve(entity)),
        latest: null,
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          operation: 'DELETE',
          property: 'missing',
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ namespace, name, latest: entity.latest }));
      expect(entity.update).toHaveBeenCalledWith(req.user, expect.objectContaining({}));
      expect(eventServiceMock.send).toHaveBeenCalledTimes(1);
    });

    it('can handle missing property on delete', async () => {
      const handler = patchConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        update: jest.fn(() => Promise.resolve(entity)),
        latest: { revision: 1, configuration: { old: 'old' } },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          operation: 'DELETE',
          property: 'missing',
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ namespace, name, latest: entity.latest }));
      expect(entity.update).toHaveBeenCalledWith(req.user, expect.objectContaining(entity.latest.configuration));
      expect(eventServiceMock.send).toHaveBeenCalledTimes(1);
    });

    it('can return error for delete without property', async () => {
      const handler = patchConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        update: jest.fn(() => Promise.resolve(entity)),
        latest: { revision: 1, configuration: { old: 'old' } },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          operation: 'DELETE',
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(InvalidOperationError));
      expect(eventServiceMock.send).toHaveBeenCalledTimes(0);
    });

    it('can return error for unrecognized operation', async () => {
      const handler = patchConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        update: jest.fn(() => Promise.resolve(entity)),
        latest: { revision: 1, configuration: { old: 'old' } },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          operation: 'NOP',
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(InvalidOperationError));
      expect(eventServiceMock.send).toHaveBeenCalledTimes(0);
    });
  });

  describe('createServiceConfigurationRevision', () => {
    it('can create handler', () => {
      const handler = createConfigurationRevision(eventServiceMock);
      expect(handler).toBeTruthy();
    });

    it('can create revision', async () => {
      const handler = createConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        createRevision: jest.fn(() => Promise.resolve(entity)),
        latest: { revision: 1, configuration: {} },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          revision: true,
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ namespace, name, latest: entity.latest }));
      expect(entity.createRevision).toHaveBeenCalledTimes(1);
      expect(eventServiceMock.send).toHaveBeenCalledTimes(1);
    });

    it('can create first revision', async () => {
      const handler = createConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        createRevision: jest.fn(() => Promise.resolve(entity)),
        latest: null,
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {
          revision: true,
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ namespace, name, latest: entity.latest }));
      expect(entity.createRevision).toHaveBeenCalledTimes(1);
      expect(eventServiceMock.send).toHaveBeenCalledTimes(1);
    });

    it('can return error for unrecognized request', async () => {
      const handler = createConfigurationRevision(eventServiceMock);

      const entity = {
        tenantId,
        namespace,
        name,
        createRevision: jest.fn(() => Promise.resolve(entity)),
        latest: { revision: 1, configuration: {} },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: {},
        body: {},
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      await handler(req, (res as unknown) as Response, next);
      expect(next).toHaveBeenCalledWith(expect.any(InvalidOperationError));
    });
  });

  describe('getRevisions', () => {
    it('can create handler', () => {
      const handler = getRevisions();
      expect(handler).toBeTruthy();
    });

    it('can get revisions', async () => {
      const handler = getRevisions();

      const entity = {
        tenantId,
        namespace,
        name,
        getRevisions: jest.fn(),
        latest: { revision: 1, configuration: {} },
      };
      const req = ({
        entity,
        user: { isCore: false, roles: [ConfigurationServiceRoles.Reader], tenantId } as User,
        params: { namespace, name },
        query: { top: '12', after: '123' },
        body: {
          revision: true,
        },
      } as unknown) as Request;

      const res = {
        send: jest.fn(),
      };

      const next = jest.fn();

      const result = {};
      entity.getRevisions.mockResolvedValueOnce(result);
      await handler(req, (res as unknown) as Response, next);
      expect(res.send).toHaveBeenCalledWith(result);
      expect(entity.getRevisions).toHaveBeenCalledWith(12, '123', {});
    });
  });
});
