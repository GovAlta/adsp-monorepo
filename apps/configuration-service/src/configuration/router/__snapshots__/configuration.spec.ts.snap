// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`router configurationOperations can create revision 1`] = `
{
  "latest": {
    "configuration": {},
    "created": Any<Date>,
    "revision": 1,
  },
  "name": "test-service",
  "namespace": "platform",
}
`;

exports[`router configurationOperations can set the active revision 1`] = `
{
  "active": 2,
  "name": "test-service",
  "namespace": "platform",
  "tenantId": AdspId {
    "api": "v2",
    "namespace": "platform",
    "resource": "/tenants/test",
    "service": "tenant-service",
    "toString": [Function],
    "type": "resource",
  },
}
`;

exports[`router getActiveRevision can fallback to latest for active revision not found 1`] = `
{
  "data": {
    "a": 42,
  },
  "name": "test-service",
  "namespace": "platform",
  "revision": 12,
}
`;

exports[`router getActiveRevision can fallback to latest for no active revision 1`] = `
{
  "data": {
    "a": 42,
  },
  "name": "test-service",
  "namespace": "platform",
  "revision": 12,
}
`;

exports[`router getActiveRevision can get active revision 1`] = `
{
  "data": {
    "a": 42,
  },
  "name": "test-service",
  "namespace": "platform",
  "revision": 3,
}
`;

exports[`router getActiveRevision can get active revision at revision 0 1`] = `
{
  "data": {
    "a": 42,
  },
  "name": "test-service",
  "namespace": "platform",
  "revision": 3,
}
`;

exports[`router getRevisions can get revisions 1`] = `{}`;

exports[`router getServiceConfiguration can get configuration 1`] = `
{
  "latest": null,
  "name": "test-service",
  "namespace": "platform",
}
`;

exports[`router patchServiceConfigurationRevision can delete 1`] = `
{
  "latest": {
    "configuration": {
      "old": "old",
    },
    "lastUpdated": Any<Date>,
    "revision": 1,
  },
  "name": "test-service",
  "namespace": "platform",
}
`;

exports[`router patchServiceConfigurationRevision can delete at path 1`] = `
{
  "latest": {
    "configuration": {
      "old": {
        "nested": "old",
      },
    },
    "lastUpdated": Any<Date>,
    "revision": 1,
  },
  "name": "test-service",
  "namespace": "platform",
}
`;

exports[`router patchServiceConfigurationRevision can replace 1`] = `
{
  "latest": {
    "configuration": {
      "old": "old",
    },
    "lastUpdated": Any<Date>,
    "revision": 1,
  },
  "name": "test-service",
  "namespace": "platform",
}
`;

exports[`router patchServiceConfigurationRevision can update 1`] = `
{
  "latest": {
    "configuration": {
      "old": "old",
    },
    "lastUpdated": Any<Date>,
    "revision": 1,
  },
  "name": "test-service",
  "namespace": "platform",
}
`;
